File: packages/sdk-ts/src/sandbox.ts
import type {
  SandboxStatus,
  ExecStreamEvent,
  Artifact,
  FileEntry,
  ExecSyncResponse,
  ExecAsyncResponse,
  ForkSandboxResponse,
  GetForkTreeResponse,
  GetSandboxResponse,
  StopSandboxResponse,
  ListFilesResponse,
  RegisterArtifactsResponse,
  ListArtifactsResponse,
  CreateSessionResponse,
} from '@sandchest/contract'
import type { HttpClient } from './http.js'
import type {
  ExecOptions,
  StreamExecOptions,
  ExecResult,
  ForkOptions,
  ForkTree,
  FileOperations,
  ArtifactOperations,
  GitCloneOptions,
  GitOperations,
  FindOptions,
  ReplaceOptions,
  ReplaceResult,
  SessionManager,
  ToolOperations,
  CreateSessionOptions,
} from './types.js'
import { Session } from './session.js'
import { ExecFailedError, TimeoutError } from './errors.js'
import { parseSSE, ExecStream } from './stream.js'

const WAIT_READY_DEFAULT_TIMEOUT = 120_000
const WAIT_READY_POLL_INTERVAL = 1_000
const TEXT_ENCODER = new TextEncoder()
const TEXT_DECODER = new TextDecoder()
const VALIDATE_TAR_SCRIPT = `
import posixpath
import sys
import tarfile

archive = sys.argv[1]
bad = []

with tarfile.open(archive, 'r:gz') as tf:
    for member in tf.getmembers():
        normalized = posixpath.normpath(member.name)
        if member.name.startswith('/'):
            bad.append(f"absolute path: {member.name}")
            continue
        if normalized == '.':
            if not member.isdir():
                bad.append(f"non-directory root entry: {member.name}")
            continue
        if normalized == '..' or normalized.startswith('../'):
            bad.append(f"path traversal: {member.name}")
            continue
        if not (member.isfile() or member.isdir()):
            bad.append(f"unsupported type: {member.name} ({member.type!r})")
            continue
        if member.issym() or member.islnk():
            bad.append(f"link entry: {member.name}")
            continue

if bad:
    sys.stderr.write('Tarball contains unsafe entries: ' + ', '.join(bad[:5]))
    sys.exit(1)
`.trim()
const REPLACE_SCRIPT = `
import fnmatch
import os
import sys

MAX_FILE_SIZE = 10 * 1024 * 1024

search_file, replace_file, root = sys.argv[1], sys.argv[2], sys.argv[3]
glob_pat = sys.argv[4] if len(sys.argv) > 4 else None
with open(search_file) as f:
    search = f.read()
with open(replace_file) as f:
    repl = f.read()
real_root = os.path.realpath(root)
root_prefix = real_root if real_root.endswith(os.sep) else real_root + os.sep
count = 0

for dirpath, _, files in os.walk(root, followlinks=False):
    for name in files:
        fp = os.path.join(dirpath, name)
        if glob_pat and not fnmatch.fnmatch(name, glob_pat):
            continue
        try:
            if os.path.islink(fp):
                continue
            resolved = os.path.realpath(fp)
            if resolved != real_root and not resolved.startswith(root_prefix):
                continue
            if os.path.getsize(resolved) > MAX_FILE_SIZE:
                continue
            with open(resolved) as f:
                data = f.read()
            if search in data:
                with open(resolved, 'w') as f:
                    f.write(data.replace(search, repl))
                count += 1
                print(resolved, end='\\0')
        except (UnicodeDecodeError, PermissionError, OSError):
            pass

print(f'Replaced in {count} file(s)', file=sys.stderr)
`.trim()

function isScpStyleGitUrl(value: string): boolean {
  return /^[^@\s]+@[^:\s]+:.+$/.test(value)
}

function validateGitCloneUrl(rawUrl: string): { ok: true; url: string } | { ok: false; error: string } {
  const url = rawUrl.trim()
  if (url === '') {
    return { ok: false, error: 'Git URL must not be empty.' }
  }

  if (isScpStyleGitUrl(url)) {
    return { ok: true, url }
  }

  if (!url.includes('://')) {
    return {
      ok: false,
      error: `Invalid git URL: ${url}. Use an HTTPS URL or an SSH-style URL such as git@github.com:org/repo.git.`,
    }
  }

  let parsed: URL
  try {
    parsed = new URL(url)
  } catch {
    return { ok: false, error: `Invalid git URL: ${url}.` }
  }

  if (parsed.username || parsed.password) {
    return {
      ok: false,
      error:
        'URLs with embedded credentials (user:pass@host) are not allowed. Private-repo auth is intentionally deferred until Sandchest can inject credentials outside the guest boundary.',
    }
  }

  return { ok: true, url }
}

/**
 * A Sandchest sandbox — an isolated Firecracker microVM.
 * All operations hang off this instance. No ID passing needed.
 */
export class Sandbox {
  readonly id: string
  status: SandboxStatus
  readonly replayUrl: string
  /** @internal */
  readonly _http: HttpClient

  /** File system operations. */
  readonly fs: FileOperations

  /** Artifact operations. */
  readonly artifacts: ArtifactOperations

  /** Git operations. */
  readonly git: GitOperations

  /** Session manager. */
  readonly session: SessionManager

  /** Exec-based helper tools. */
  readonly tools: ToolOperations

  /** @internal — Use `sandchest.create()` or `sandchest.get()` instead. */
  constructor(id: string, status: SandboxStatus, replayUrl: string, http: HttpClient) {
    this.id = id
    this.status = status
    this.replayUrl = replayUrl
    this._http = http

    this.fs = {
      upload: async (path: string, content: Uint8Array): Promise<void> => {
        await this._http.requestRaw({
          method: 'PUT',
          path: `/v1/sandboxes/${this.id}/files`,
          query: { path },
          body: content,
          headers: { 'Content-Type': 'application/octet-stream' },
        })
      },
      write: async (path: string, content: string): Promise<void> => {
        await this.fs.upload(path, TEXT_ENCODER.encode(content))
      },
      uploadDir: async (path: string, tarball: Uint8Array): Promise<void> => {
        const tmpPath = `/tmp/.sandchest-upload-${crypto.randomUUID()}.tar.gz`
        await this._http.requestRaw({
          method: 'PUT',
          path: `/v1/sandboxes/${this.id}/files`,
          query: { path: tmpPath, batch: true },
          body: tarball,
          headers: { 'Content-Type': 'application/octet-stream' },
        })
        try {
          const mkdirResult = await this._execSync(['mkdir', '-p', path], {
            operation: 'uploadDir:mkdir',
          })
          if (mkdirResult.exit_code !== 0) {
            throw new ExecFailedError({
              operation: 'uploadDir:mkdir',
              exitCode: mkdirResult.exit_code,
              stderr: mkdirResult.stderr,
            })
          }

          const validateResult = await this._execSync([
            'python3',
            '-c',
            VALIDATE_TAR_SCRIPT,
            tmpPath,
          ], {
            operation: 'uploadDir:verify',
          })
          if (validateResult.exit_code !== 0) {
            throw new ExecFailedError({
              operation: 'uploadDir:verify',
              exitCode: validateResult.exit_code,
              stderr: validateResult.stderr,
            })
          }

          const extractResult = await this._execSync([
            'tar',
            'xzf',
            tmpPath,
            '--no-same-owner',
            '-C',
            path,
          ], {
            operation: 'uploadDir:extract',
          })
          if (extractResult.exit_code !== 0) {
            throw new ExecFailedError({
              operation: 'uploadDir:extract',
              exitCode: extractResult.exit_code,
              stderr: extractResult.stderr,
            })
          }
        } finally {
          await this._execSync(['rm', '-f', tmpPath], {
            operation: 'uploadDir:cleanup',
          }).catch(() => {})
        }
      },
      download: async (path: string): Promise<Uint8Array> => {
        const res = await this._http.requestRaw({
          method: 'GET',
          path: `/v1/sandboxes/${this.id}/files`,
          query: { path },
        })
        return new Uint8Array(await res.arrayBuffer())
      },
      read: async (path: string): Promise<string> => {
        return TEXT_DECODER.decode(await this.fs.download(path))
      },
      downloadDir: async (path: string): Promise<Uint8Array> => {
        const tmpPath = `/tmp/.sandchest-download-${crypto.randomUUID()}.tar.gz`
        try {
          const archiveResult = await this._execSync(['tar', 'czf', tmpPath, '-C', path, '.'], {
            operation: 'downloadDir:archive',
          })
          if (archiveResult.exit_code !== 0) {
            throw new ExecFailedError({
              operation: 'downloadDir:archive',
              exitCode: archiveResult.exit_code,
              stderr: archiveResult.stderr,
            })
          }

          return await this.fs.download(tmpPath)
        } finally {
          await this._execSync(['rm', '-f', tmpPath], {
            operation: 'downloadDir:cleanup',
          }).catch(() => {})
        }
      },
      ls: async (path: string): Promise<FileEntry[]> => {
        const res = await this._http.request<ListFilesResponse>({
          method: 'GET',
          path: `/v1/sandboxes/${this.id}/files`,
          query: { path, list: true },
        })
        return res.files
      },
      rm: async (path: string): Promise<void> => {
        await this._http.request<{ ok: true }>({
          method: 'DELETE',
          path: `/v1/sandboxes/${this.id}/files`,
          query: { path },
        })
      },
    }

    this.artifacts = {
      register: async (paths: string[]): Promise<{ registered: number; total: number }> => {
        const res = await this._http.request<RegisterArtifactsResponse>({
          method: 'POST',
          path: `/v1/sandboxes/${this.id}/artifacts`,
          body: { paths },
        })
        return { registered: res.registered, total: res.total }
      },
      list: async (): Promise<Artifact[]> => {
        const res = await this._http.request<ListArtifactsResponse>({
          method: 'GET',
          path: `/v1/sandboxes/${this.id}/artifacts`,
        })
        return res.artifacts
      },
    }

    this.git = {
      clone: async (url: string, options?: GitCloneOptions): Promise<ExecResult> => {
        const validated = validateGitCloneUrl(url)
        if (!validated.ok) {
          throw new ExecFailedError({
            operation: 'git.clone',
            exitCode: 1,
            stderr: validated.error,
          })
        }

        const dest = options?.dest ?? '/work'
        const cmd: string[] = ['git', 'clone']

        if (options?.branch) {
          if (options.branch.startsWith('-')) {
            throw new ExecFailedError({
              operation: 'git.clone',
              exitCode: 1,
              stderr: `Invalid branch name: '${options.branch}' — branch names must not start with '-'`,
            })
          }
          cmd.push('--branch', options.branch)
        }

        if (options?.depth !== undefined) {
          cmd.push('--depth', String(options.depth))
        }

        if (options?.singleBranch !== false) {
          cmd.push('--single-branch')
        }

        cmd.push('--', validated.url, dest)

        const result = await this._execSync(cmd, {
          operation: 'git.clone',
          timeout: options?.timeout ?? 120,
          env: {
            GIT_TERMINAL_PROMPT: '0',
            ...options?.env,
          },
        })

        if (result.exit_code !== 0) {
          throw new ExecFailedError({
            operation: 'git.clone',
            exitCode: result.exit_code,
            stderr: result.stderr,
          })
        }

        return {
          execId: result.exec_id,
          exitCode: result.exit_code,
          stdout: result.stdout,
          stderr: result.stderr,
          durationMs: result.duration_ms,
        }
      },
    }

    this.session = {
      create: async (options?: CreateSessionOptions): Promise<Session> => {
        const res = await this._http.request<CreateSessionResponse>({
          method: 'POST',
          path: `/v1/sandboxes/${this.id}/sessions`,
          body: {
            shell: options?.shell,
            env: options?.env,
          },
        })
        return new Session(res.session_id, this.id, this._http)
      },
    }

    this.tools = {
      find: async (path: string, pattern: string, options?: FindOptions): Promise<string[]> => {
        const cmd = ['find', '--', path]
        if (options?.maxDepth !== undefined) {
          cmd.push('-maxdepth', String(options.maxDepth))
        }
        if (options?.type) {
          cmd.push('-type', options.type)
        }
        cmd.push('-name', pattern)

        const result = await this._execSync(cmd, { operation: 'tools.find' })
        if (result.exit_code !== 0) {
          return []
        }
        return result.stdout.trim().split('\n').filter(Boolean)
      },
      replace: async (
        path: string,
        search: string,
        replacement: string,
        options?: ReplaceOptions,
      ): Promise<ReplaceResult> => {
        if (search.length === 0) {
          throw new ExecFailedError({
            operation: 'replace',
            exitCode: 1,
            stderr: 'search string must not be empty',
          })
        }

        const id = crypto.randomUUID()
        const searchPath = `/tmp/.sandchest-search-${id}`
        const replacePath = `/tmp/.sandchest-replace-${id}`

        try {
          await this.fs.upload(searchPath, TEXT_ENCODER.encode(search))
          await this.fs.upload(replacePath, TEXT_ENCODER.encode(replacement))

          const result = await this._execSync(
            [
              'python3',
              '-c',
              REPLACE_SCRIPT,
              searchPath,
              replacePath,
              path,
              ...(options?.glob ? [options.glob] : []),
            ],
            { operation: 'replace' },
          )

          if (result.exit_code !== 0) {
            throw new ExecFailedError({
              operation: 'replace',
              exitCode: result.exit_code,
              stderr: result.stderr,
            })
          }

          const changedPaths = result.stdout.split('\0').filter(Boolean)
          return {
            filesChanged: changedPaths.length,
            changedPaths,
          }
        } finally {
          for (const tempPath of [searchPath, replacePath]) {
            await this._execSync(['rm', '-f', tempPath], {
              operation: 'replace:cleanup',
            }).catch(() => {})
          }
        }
      },
    }
  }

  /** Execute a command (blocking, returns result). */
  exec(cmd: string | string[], options?: ExecOptions): Promise<ExecResult>
  /** Execute a command (streaming, returns ExecStream). */
  exec(cmd: string | string[], options: StreamExecOptions): Promise<ExecStream>
  exec(
    cmd: string | string[],
    options?: ExecOptions | StreamExecOptions,
  ): Promise<ExecResult> | Promise<ExecStream> {
    if (options && 'stream' in options && options.stream === true) {
      return this._execStream(cmd, options)
    }
    return this._execBlocking(cmd, options as ExecOptions | undefined)
  }

  /** Fork this sandbox's entire state into a new sandbox. */
  async fork(options?: ForkOptions): Promise<Sandbox> {
    const res = await this._http.request<ForkSandboxResponse>({
      method: 'POST',
      path: `/v1/sandboxes/${this.id}/fork`,
      body: {
        env: options?.env,
        ttl_seconds: options?.ttlSeconds,
      },
    })
    return new Sandbox(res.sandbox_id, res.status, res.replay_url, this._http)
  }

  /** Get the fork tree rooted at this sandbox. */
  async forks(): Promise<ForkTree> {
    const res = await this._http.request<GetForkTreeResponse>({
      method: 'GET',
      path: `/v1/sandboxes/${this.id}/forks`,
    })
    return { root: res.root, tree: res.tree }
  }

  /** Gracefully stop this sandbox (collects artifacts). */
  async stop(): Promise<void> {
    const res = await this._http.request<StopSandboxResponse>({
      method: 'POST',
      path: `/v1/sandboxes/${this.id}/stop`,
    })
    this.status = res.status
  }

  /** Hard stop and clean up this sandbox. */
  async destroy(): Promise<void> {
    await this._http.request<{ sandbox_id: string; status: 'deleted' }>({
      method: 'DELETE',
      path: `/v1/sandboxes/${this.id}`,
    })
    this.status = 'deleted'
  }

  /** Wait for this sandbox to reach 'running' status. */
  async waitReady(options?: { timeout?: number | undefined }): Promise<void> {
    const timeout = options?.timeout ?? WAIT_READY_DEFAULT_TIMEOUT
    const start = Date.now()

    while (true) {
      const res = await this._http.request<GetSandboxResponse>({
        method: 'GET',
        path: `/v1/sandboxes/${this.id}`,
      })

      this.status = res.status

      if (res.status === 'running') return

      if (res.status === 'failed' || res.status === 'deleted' || res.status === 'stopped') {
        throw new Error(`Sandbox ${this.id} reached terminal state: ${res.status}`)
      }

      if (Date.now() - start >= timeout) {
        throw new TimeoutError({
          message: `Sandbox ${this.id} did not become ready within ${timeout}ms`,
          timeoutMs: timeout,
        })
      }

      await new Promise((resolve) => setTimeout(resolve, WAIT_READY_POLL_INTERVAL))
    }
  }

  /** Auto-cleanup via Explicit Resource Management. Calls stop() if running. */
  async [Symbol.asyncDispose](): Promise<void> {
    if (this.status === 'running') {
      await this.stop()
    }
  }

  private async _execBlocking(
    cmd: string | string[],
    options?: ExecOptions,
  ): Promise<ExecResult> {
    if (options?.onStdout || options?.onStderr) {
      return this._execWithCallbacks(cmd, options)
    }

    const res = await this._http.request<ExecSyncResponse>({
      method: 'POST',
      path: `/v1/sandboxes/${this.id}/exec`,
      body: {
        cmd,
        cwd: options?.cwd,
        env: options?.env,
        timeout_seconds: options?.timeout,
        wait: true,
      },
    })

    return {
      execId: res.exec_id,
      exitCode: res.exit_code,
      stdout: res.stdout,
      stderr: res.stderr,
      durationMs: res.duration_ms,
    }
  }

  private async _execSync(
    cmd: string[],
    options?: {
      operation?: string | undefined
      timeout?: number | undefined
      env?: Record<string, string> | undefined
    },
  ): Promise<ExecSyncResponse> {
    const result = await this._http.request<ExecSyncResponse>({
      method: 'POST',
      path: `/v1/sandboxes/${this.id}/exec`,
      body: {
        cmd,
        env: options?.env,
        timeout_seconds: options?.timeout ?? 30,
        wait: true,
      },
    })

    if (result.status !== 'done') {
      throw new ExecFailedError({
        operation: options?.operation ?? 'exec',
        exitCode: result.exit_code,
        stderr: result.stderr || `Command failed with status: ${result.status}`,
      })
    }

    return result
  }

  private async _execWithCallbacks(
    cmd: string | string[],
    options: ExecOptions,
  ): Promise<ExecResult> {
    const asyncRes = await this._http.request<ExecAsyncResponse>({
      method: 'POST',
      path: `/v1/sandboxes/${this.id}/exec`,
      body: {
        cmd,
        cwd: options.cwd,
        env: options.env,
        timeout_seconds: options.timeout,
        wait: false,
      },
    })

    const response = await this._http.requestRaw({
      method: 'GET',
      path: `/v1/sandboxes/${this.id}/exec/${asyncRes.exec_id}/stream`,
      headers: { Accept: 'text/event-stream' },
    })

    let stdout = ''
    let stderr = ''
    let exitCode = 0
    let durationMs = 0

    for await (const event of parseSSE<ExecStreamEvent>(response)) {
      switch (event.t) {
        case 'stdout':
          stdout += event.data
          options.onStdout?.(event.data)
          break
        case 'stderr':
          stderr += event.data
          options.onStderr?.(event.data)
          break
        case 'exit':
          exitCode = event.code
          durationMs = event.duration_ms
          break
      }
    }

    return { execId: asyncRes.exec_id, exitCode, stdout, stderr, durationMs }
  }

  private async _execStream(
    cmd: string | string[],
    options: StreamExecOptions,
  ): Promise<ExecStream> {
    const asyncRes = await this._http.request<ExecAsyncResponse>({
      method: 'POST',
      path: `/v1/sandboxes/${this.id}/exec`,
      body: {
        cmd,
        cwd: options.cwd,
        env: options.env,
        timeout_seconds: options.timeout,
        wait: false,
      },
    })

    const response = await this._http.requestRaw({
      method: 'GET',
      path: `/v1/sandboxes/${this.id}/exec/${asyncRes.exec_id}/stream`,
      headers: { Accept: 'text/event-stream' },
    })

    return new ExecStream(asyncRes.exec_id, parseSSE<ExecStreamEvent>(response))
  }
}

File: packages/sdk-py/sandchest/sandbox.py
"""Sandbox — an isolated Firecracker microVM."""

from __future__ import annotations

import time
from collections.abc import Callable
from typing import TYPE_CHECKING

from .errors import TimeoutError
from .session import Session
from .stream import ExecStream, parse_sse
from .types import (
    Artifact,
    ExecResult,
    FileEntry,
    ForkTree,
    ForkTreeNode,
    RegisterArtifactsResult,
    SandboxStatus,
)

if TYPE_CHECKING:
    from .http import HttpClient

WAIT_READY_DEFAULT_TIMEOUT_MS = 120_000
WAIT_READY_POLL_INTERVAL_S = 1.0


class Sandbox:
    """A Sandchest sandbox — an isolated Firecracker microVM.

    All operations hang off this instance. Use ``Sandchest.create()``
    or ``Sandchest.get()`` to obtain an instance.
    """

    def __init__(
        self,
        id: str,
        status: SandboxStatus,
        replay_url: str,
        http: HttpClient,
    ) -> None:
        self.id = id
        self.status: SandboxStatus = status
        self.replay_url = replay_url
        self._http = http

    # -- Context manager support -----------------------------------------------

    def __enter__(self) -> Sandbox:
        return self

    def __exit__(self, *_: object) -> None:
        if self.status == "running":
            self.stop()

    # -- Execution -------------------------------------------------------------

    def exec(
        self,
        cmd: str | list[str],
        *,
        cwd: str | None = None,
        env: dict[str, str] | None = None,
        timeout: int | None = None,
        stream: bool = False,
        on_stdout: Callable[[str], None] | None = None,
        on_stderr: Callable[[str], None] | None = None,
    ) -> ExecResult | ExecStream:
        """Execute a command in the sandbox.

        Args:
            cmd: Command string or list of args.
            cwd: Working directory.
            env: Environment variables.
            timeout: Timeout in seconds.
            stream: If True, return an ExecStream for event-by-event iteration.
            on_stdout: Callback for each stdout chunk (triggers streaming internally).
            on_stderr: Callback for each stderr chunk (triggers streaming internally).

        Returns:
            ExecResult when stream=False (default), ExecStream when stream=True.
        """
        if stream:
            return self._exec_stream(cmd, cwd=cwd, env=env, timeout=timeout)
        if on_stdout or on_stderr:
            return self._exec_with_callbacks(
                cmd,
                cwd=cwd,
                env=env,
                timeout=timeout,
                on_stdout=on_stdout,
                on_stderr=on_stderr,
            )
        return self._exec_blocking(cmd, cwd=cwd, env=env, timeout=timeout)

    # -- Lifecycle -------------------------------------------------------------

    def fork(
        self,
        *,
        env: dict[str, str] | None = None,
        ttl_seconds: int | None = None,
    ) -> Sandbox:
        """Fork this sandbox's entire state into a new sandbox."""
        res = self._http.request(
            "POST",
            f"/v1/sandboxes/{self.id}/fork",
            body={
                "env": env,
                "ttl_seconds": ttl_seconds,
            },
        )
        return Sandbox(res["sandbox_id"], res["status"], res["replay_url"], self._http)

    def forks(self) -> ForkTree:
        """Get the fork tree rooted at this sandbox."""
        res = self._http.request(
            "GET",
            f"/v1/sandboxes/{self.id}/forks",
        )
        return ForkTree(
            root=res["root"],
            tree=[
                ForkTreeNode(
                    sandbox_id=n["sandbox_id"],
                    status=n["status"],
                    forked_from=n.get("forked_from"),
                    forked_at=n.get("forked_at"),
                    children=n.get("children", []),
                )
                for n in res["tree"]
            ],
        )

    def stop(self) -> None:
        """Gracefully stop this sandbox (collects artifacts)."""
        res = self._http.request(
            "POST",
            f"/v1/sandboxes/{self.id}/stop",
        )
        self.status = res["status"]

    def destroy(self) -> None:
        """Hard stop and clean up this sandbox."""
        self._http.request(
            "DELETE",
            f"/v1/sandboxes/{self.id}",
        )
        self.status = "deleted"

    def wait_ready(self, *, timeout: int | None = None) -> None:
        """Wait for this sandbox to reach 'running' status.

        Args:
            timeout: Max wait time in milliseconds. Defaults to 120000.
        """
        timeout_ms = timeout if timeout is not None else WAIT_READY_DEFAULT_TIMEOUT_MS
        start = time.monotonic()

        while True:
            res = self._http.request(
                "GET",
                f"/v1/sandboxes/{self.id}",
            )
            self.status = res["status"]

            if res["status"] == "running":
                return

            if res["status"] in ("failed", "deleted", "stopped"):
                raise RuntimeError(
                    f"Sandbox {self.id} reached terminal state: {res['status']}"
                )

            elapsed_ms = (time.monotonic() - start) * 1000
            if elapsed_ms >= timeout_ms:
                msg = (
                    f"Sandbox {self.id} did not become "
                    f"ready within {timeout_ms}ms"
                )
                raise TimeoutError(
                    message=msg,
                    timeout_ms=timeout_ms,
                )

            time.sleep(WAIT_READY_POLL_INTERVAL_S)

    # -- File operations -------------------------------------------------------

    def upload(self, path: str, content: bytes) -> None:
        """Upload a file to the sandbox."""
        self._http.request_raw(
            "PUT",
            f"/v1/sandboxes/{self.id}/files",
            query={"path": path},
            body=content,
            headers={"Content-Type": "application/octet-stream"},
        )

    def upload_dir(self, path: str, tarball: bytes) -> None:
        """Upload a directory (as tarball) to the sandbox."""
        self._http.request_raw(
            "PUT",
            f"/v1/sandboxes/{self.id}/files",
            query={"path": path, "batch": "true"},
            body=tarball,
            headers={"Content-Type": "application/octet-stream"},
        )

    def download(self, path: str) -> bytes:
        """Download a file from the sandbox."""
        response = self._http.request_raw(
            "GET",
            f"/v1/sandboxes/{self.id}/files",
            query={"path": path},
        )
        return response.content

    def ls(self, path: str) -> list[FileEntry]:
        """List directory contents."""
        res = self._http.request(
            "GET",
            f"/v1/sandboxes/{self.id}/files",
            query={"path": path, "list": "true"},
        )
        return [
            FileEntry(
                name=f["name"],
                path=f["path"],
                type=f["type"],
                size_bytes=f.get("size_bytes"),
            )
            for f in res["files"]
        ]

    def rm(self, path: str) -> None:
        """Delete a file from the sandbox."""
        self._http.request(
            "DELETE",
            f"/v1/sandboxes/{self.id}/files",
            query={"path": path},
        )

    # -- Artifact operations ---------------------------------------------------

    def register_artifacts(self, paths: list[str]) -> RegisterArtifactsResult:
        """Register files as artifacts for collection on stop."""
        res = self._http.request(
            "POST",
            f"/v1/sandboxes/{self.id}/artifacts",
            body={"paths": paths},
        )
        return RegisterArtifactsResult(
            registered=res["registered"], total=res["total"]
        )

    def list_artifacts(self) -> list[Artifact]:
        """List artifacts registered in this sandbox."""
        res = self._http.request(
            "GET",
            f"/v1/sandboxes/{self.id}/artifacts",
        )
        return [
            Artifact(
                id=a["id"],
                name=a["name"],
                mime=a["mime"],
                bytes=a["bytes"],
                sha256=a["sha256"],
                download_url=a["download_url"],
                exec_id=a.get("exec_id"),
                created_at=a["created_at"],
            )
            for a in res["artifacts"]
        ]

    # -- Session management ----------------------------------------------------

    def create_session(
        self,
        *,
        shell: str | None = None,
        env: dict[str, str] | None = None,
    ) -> Session:
        """Create a stateful session in this sandbox."""
        res = self._http.request(
            "POST",
            f"/v1/sandboxes/{self.id}/sessions",
            body={
                "shell": shell,
                "env": env,
            },
        )
        return Session(res["session_id"], self.id, self._http)

    # -- Private execution helpers ---------------------------------------------

    def _exec_blocking(
        self,
        cmd: str | list[str],
        *,
        cwd: str | None = None,
        env: dict[str, str] | None = None,
        timeout: int | None = None,
    ) -> ExecResult:
        res = self._http.request(
            "POST",
            f"/v1/sandboxes/{self.id}/exec",
            body={
                "cmd": cmd,
                "cwd": cwd,
                "env": env,
                "timeout_seconds": timeout,
                "wait": True,
            },
        )
        return ExecResult(
            exec_id=res["exec_id"],
            exit_code=res["exit_code"],
            stdout=res["stdout"],
            stderr=res["stderr"],
            duration_ms=res["duration_ms"],
        )

    def _exec_with_callbacks(
        self,
        cmd: str | list[str],
        *,
        cwd: str | None = None,
        env: dict[str, str] | None = None,
        timeout: int | None = None,
        on_stdout: Callable[[str], None] | None = None,
        on_stderr: Callable[[str], None] | None = None,
    ) -> ExecResult:
        async_res = self._http.request(
            "POST",
            f"/v1/sandboxes/{self.id}/exec",
            body={
                "cmd": cmd,
                "cwd": cwd,
                "env": env,
                "timeout_seconds": timeout,
                "wait": False,
            },
        )

        response = self._http.request_stream(
            "GET",
            f"/v1/sandboxes/{self.id}/exec/{async_res['exec_id']}/stream",
            headers={"Accept": "text/event-stream"},
        )

        stdout = ""
        stderr = ""
        exit_code = 0
        duration_ms = 0

        for event in parse_sse(response):
            if event["t"] == "stdout":
                stdout += event["data"]
                if on_stdout:
                    on_stdout(event["data"])
            elif event["t"] == "stderr":
                stderr += event["data"]
                if on_stderr:
                    on_stderr(event["data"])
            elif event["t"] == "exit":
                exit_code = event["code"]
                duration_ms = event["duration_ms"]

        return ExecResult(
            exec_id=async_res["exec_id"],
            exit_code=exit_code,
            stdout=stdout,
            stderr=stderr,
            duration_ms=duration_ms,
        )

    def _exec_stream(
        self,
        cmd: str | list[str],
        *,
        cwd: str | None = None,
        env: dict[str, str] | None = None,
        timeout: int | None = None,
    ) -> ExecStream:
        async_res = self._http.request(
            "POST",
            f"/v1/sandboxes/{self.id}/exec",
            body={
                "cmd": cmd,
                "cwd": cwd,
                "env": env,
                "timeout_seconds": timeout,
                "wait": False,
            },
        )

        response = self._http.request_stream(
            "GET",
            f"/v1/sandboxes/{self.id}/exec/{async_res['exec_id']}/stream",
            headers={"Accept": "text/event-stream"},
        )

        return ExecStream(async_res["exec_id"], parse_sse(response))
